Create a zone for each cluster domain

njuettner opened the pull request:
Issue: https://github.com/giantswarm/giantswarm/issues/27506

When having multiple cluster domains, we should generate automatically a CoreDNS for every cluster domain.

E.g. when we have `cluster.local` and `my.zone.com` in `clusterDomain` value, we should separate them. This is the case for some clusters using Linkerd with a different trust domain. It is helpful for analyzing in case we see errors, because CoreDNS exports metrics by zones.

The PR also removes the kubernetes plugin from `.` zone and adds it to the `*.local` which is coming from `clusterDomain` value.

**Example 1:**

```bash 
helm template helm/coredns-app/ --set cluster.kubernetes.clusterDomain="cluster.local"
```

Output: 
```yaml
---
# Source: coredns-app/templates/configmap.yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: coredns
  namespace: kube-system
  labels:
    k8s-app: "coredns"
    app.kubernetes.io/managed-by: "Helm"
    app.kubernetes.io/name: "coredns"
    app.kubernetes.io/instance: "release-name"
    app.kubernetes.io/version: "1.9.3"
    giantswarm.io/service-type: "managed"
    helm.sh/chart: "coredns-app-1.9.3"
    kubernetes.io/cluster-service: "true"
    kubernetes.io/name: "CoreDNS"
data:
  Corefile: |
    .:1053 {
        cache 30
        errors
        template ANY AAAA {
          rcode NOERROR
        }
        health {
          lameduck 5s
        }
        ready
        log . {
          class denial
          class error
        }
        loadbalance round_robin
        prometheus :9153
        forward . /etc/resolv.conf
        reload
    }
    cluster.local:1053 {
        errors
        template ANY AAAA {
          rcode NOERROR
        }
        health {
          lameduck 5s
        }
        ready
        kubernetes cluster.local 172.31.0.0/16 192.168.0.0/16 {
          fallthrough in-addr.arpa ip6.arpa
          pods verified
        }
        log . {
          class denial
          class error
        }
        loadbalance round_robin
        prometheus :9153
        reload
    }
---
```

**Example 2:**

```bash
helm template helm/coredns-app/ --set cluster.kubernetes.clusterDomain="my.zone.com cluster.local" 
```

Output:

```yaml
---
# Source: coredns-app/templates/configmap.yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: coredns
  namespace: kube-system
  labels:
    k8s-app: "coredns"
    app.kubernetes.io/managed-by: "Helm"
    app.kubernetes.io/name: "coredns"
    app.kubernetes.io/instance: "release-name"
    app.kubernetes.io/version: "1.9.3"
    giantswarm.io/service-type: "managed"
    helm.sh/chart: "coredns-app-1.9.3"
    kubernetes.io/cluster-service: "true"
    kubernetes.io/name: "CoreDNS"
data:
  Corefile: |
    .:1053 {
        cache 30
        errors
        template ANY AAAA {
          rcode NOERROR
        }
        health {
          lameduck 5s
        }
        ready
        log . {
          class denial
          class error
        }
        loadbalance round_robin
        prometheus :9153
        forward . /etc/resolv.conf
        reload
    }
    my.zone.com:1053 {
        errors
        template ANY AAAA {
          rcode NOERROR
        }
        health {
          lameduck 5s
        }
        ready
        kubernetes my.zone.com {
          fallthrough in-addr.arpa ip6.arpa
          pods verified
        }
        log . {
          class denial
          class error
        }
        loadbalance round_robin
        prometheus :9153
        reload
    }
    cluster.local:1053 {
        errors
        template ANY AAAA {
          rcode NOERROR
        }
        health {
          lameduck 5s
        }
        ready
        kubernetes cluster.local 172.31.0.0/16 192.168.0.0/16 {
          fallthrough in-addr.arpa ip6.arpa
          pods verified
        }
        log . {
          class denial
          class error
        }
        loadbalance round_robin
        prometheus :9153
        reload
    }
---
```

mcharriere commented on helm/coredns-app/templates/configmap.yaml:
```diff
@@ -46,6 +42,38 @@ data:
         {{- end }}
     }
 
+    {{- $domains := split " " .Values.cluster.kubernetes.clusterDomain }}
```
I'd really like to see a list here instead, but it might complicate the upgrade. not sure if it's worth to do some helm magic to make it work. otherwise looks good.

njuettner replied:
You're right, let me create a follow-up issue. Currently I don't know how to do that without breaking upgrades

whites11 reviewed (approved):
love it! I agree with Mati it would be best to have a separate field for the additional domains, but we would break customer configs so it's best to keep the "split" approach.

njuettner commented:
Ok I'll merge it for now and we can create a follow-up issue for that. I really don't know yet how to do it without breaking customer config.

Gacko commented:
Well, you could implement the new field right now and just default on your string. Check if the list of domains is empty, if so, fall back to the string. Then you could already add a hint to the release notes, stating that there's a new field customers can move on to while the string is being deprecated.

njuettner commented:
The thing is string `clusterDomain` cannot be deprecated, because the customer can override it via configuration and it is also used my multiple other operators. 

The `clusterDomain` value was never intended as multiple domain but it is been misused by customer which uses linkerd since forever, e.g. `cluster.local my.linkerd.company.com` as one string. 

We could add the field `additionalLocalZones` as a list, which can be picked up by customer in the future which adds the additional local zone (e.g. for linkerd), but we would also need to tell customer to remove `my.linkerd.company.com` and add it to `additionalLocalZones` when upgrading at some point.

njuettner commented:
@mcharriere @Gacko 

I changed it again, I added a new value `additionalLocalZones` which can be uncommented

It will not add any additional zone in case it is already in the `clusterDomain` string, however it will add new ones.

**Example 1:** 

```bash
helm template helm/coredns-app/ --set cluster.kubernetes.clusterDomain="my.zone.com cluster.local" --set "additionalLocalZones={my.zone.com,my.zone2.com,foo.bar}"
```

```
data:
  Corefile: |
    .:1053 {
        cache 30
        errors
        template ANY AAAA {
          rcode NOERROR
        }
        health {
          lameduck 5s
        }
        ready
        log . {
          class denial
          class error
        }
        loadbalance round_robin
        prometheus :9153
        forward . /etc/resolv.conf
        reload
    }
    cluster.local:1053 {
        errors
        template ANY AAAA {
          rcode NOERROR
        }
        health {
          lameduck 5s
        }
        ready
        kubernetes cluster.local 172.31.0.0/16 192.168.0.0/16 {
          fallthrough in-addr.arpa ip6.arpa
          pods verified
        }
        log . {
          class denial
          class error
        }
        loadbalance round_robin
        prometheus :9153
        reload
    }
    my.zone.com:1053 {
        errors
        template ANY AAAA {
          rcode NOERROR
        }
        health {
          lameduck 5s
        }
        ready
        kubernetes my.zone.com {
          fallthrough in-addr.arpa ip6.arpa
          pods verified
        }
        log . {
          class denial
          class error
        }
        loadbalance round_robin
        prometheus :9153
        reload
    }
    my.zone2.com:1053 {
        errors
        template ANY AAAA {
          rcode NOERROR
        }
        health {
          lameduck 5s
        }
        ready
        kubernetes my.zone2.com {
          fallthrough in-addr.arpa ip6.arpa
          pods verified
        }
        log . {
          class denial
          class error
        }
        loadbalance round_robin
        prometheus :9153
        reload
    }
    foo.bar:1053 {
        errors
        template ANY AAAA {
          rcode NOERROR
        }
        health {
          lameduck 5s
        }
        ready
        kubernetes foo.bar {
          fallthrough in-addr.arpa ip6.arpa
          pods verified
        }
        log . {
          class denial
          class error
        }
        loadbalance round_robin
        prometheus :9153
        reload
    }
```

**Example 2:**

```bash
helm template helm/coredns-app/ --set cluster.kubernetes.clusterDomain="my.zone.com cluster.local"
```

```yaml
data:
  Corefile: |
    .:1053 {
        cache 30
        errors
        template ANY AAAA {
          rcode NOERROR
        }
        health {
          lameduck 5s
        }
        ready
        log . {
          class denial
          class error
        }
        loadbalance round_robin
        prometheus :9153
        forward . /etc/resolv.conf
        reload
    }
    my.zone.com:1053 {
        errors
        template ANY AAAA {
          rcode NOERROR
        }
        health {
          lameduck 5s
        }
        ready
        kubernetes my.zone.com {
          fallthrough in-addr.arpa ip6.arpa
          pods verified
        }
        log . {
          class denial
          class error
        }
        loadbalance round_robin
        prometheus :9153
        reload
    }
    cluster.local:1053 {
        errors
        template ANY AAAA {
          rcode NOERROR
        }
        health {
          lameduck 5s
        }
        ready
        kubernetes cluster.local 172.31.0.0/16 192.168.0.0/16 {
          fallthrough in-addr.arpa ip6.arpa
          pods verified
        }
        log . {
          class denial
          class error
        }
        loadbalance round_robin
        prometheus :9153
        reload
    }
```

mcharriere commented on helm/coredns-app/templates/configmap.yaml:
```diff
@@ -46,6 +42,70 @@ data:
+    {{- $additionalLocalZones := .Values.additionalLocalZones -}}
+    {{- range $index, $domain := $domains }}
+    {{- if not (has $domain $additionalLocalZones) }}
+    {{ $domain }}:{{ $dnsPort }} {
```
we need to add `{{$clusterIPRange}}:{{ $dnsPort }}` and `{{ $calicoCIDR }}:{{$dnsPort}}` here, otherwise this breaks PTRs

njuettner replied:
I changed it, so now we have

```bash
helm template helm/coredns-app/ --set cluster.kubernetes.clusterDomain="my.zone.com cluster.local" --set "additionalLocalZones={my.zone.com}"
```

```yaml
data:
  Corefile: |
    .:1053 {
        cache 30
        errors
        template ANY AAAA {
          rcode NOERROR
        }
        health {
          lameduck 5s
        }
        ready
        log . {
          class denial
          class error
        }
        loadbalance round_robin
        prometheus :9153
        forward . /etc/resolv.conf
        reload
    }
    cluster.local:1053 172.31.0.0/16:1053 192.168.0.0/16:1053 {
        errors
        template ANY AAAA {
          rcode NOERROR
        }
        health {
          lameduck 5s
        }
        ready
        kubernetes cluster.local 172.31.0.0/16 192.168.0.0/16 {
          fallthrough in-addr.arpa ip6.arpa
          pods verified
        }
        log . {
          class denial
          class error
        }
        loadbalance round_robin
        prometheus :9153
        reload
    }
    my.zone.com:1053 {
        errors
        template ANY AAAA {
          rcode NOERROR
        }
        health {
          lameduck 5s
        }
        ready
        kubernetes my.zone.com {
          fallthrough in-addr.arpa ip6.arpa
          pods verified
        }
        log . {
          class denial
          class error
        }
        loadbalance round_robin
        prometheus :9153
        reload
    }
```

mcharriere reviewed (approved):
looks good!

calvix commented:
this actually is broken for CAPi as the loading of the values is too vintage-specific
```
 {{- $clusterIPRange := .Values.cluster.kubernetes.API.clusterIPRange }}
 {{- $calicoCIDR:= .Values.cluster.calico.CIDR }}
```

resulting confgiamp for coredns is
```
    cluster.local:1053 :1053 :1053 {
        errors
        template ANY AAAA {
          rcode NOERROR
        }
```
which is invalid and coredns start crashlooping
